#### Author: Diogo Viana

#### Data Scientist & Data Analyst

2024

Contact information
- https://github.com/DiogocViana\n
- https://www.linkedin.com/in/diogo-viana/

This dataset, acquired from Kaggle (https://www.kaggle.com/datasets/dillonmyrick/bike-store-sample-database/data), represents a sample dataset for a retail bike store sourced from sqlservertutorial.net. Within this notebook, I aim to showcase my proficiency in working with SQL datasets and querying relational databases using Python. Please note that this notebook is a work in progress, and I anticipate refining this introduction section in subsequent iterations

In [1]:
import pandas as pd
import sqlite3

In [5]:
#After downloading the relational database, import all its tables

brands = pd.read_csv('brands.csv')
categories = pd.read_csv('categories.csv')
customers = pd.read_csv('customers.csv')
order_items = pd.read_csv('order_items.csv')
orders = pd.read_csv('orders.csv')
products = pd.read_csv('products.csv')
staffs = pd.read_csv('staffs.csv')
stocks = pd.read_csv('stocks.csv')
stores = pd.read_csv('stores.csv')

In [6]:
#Create the database conection
connection = sqlite3.connect('bike_store.db')

# Insert data into database
brands.to_sql('brands', connection, if_exists='replace', index=False)
categories.to_sql('categories', connection, if_exists='replace', index=False)
customers.to_sql('customers', connection, if_exists='replace', index=False)
order_items.to_sql('order_items', connection, if_exists='replace', index=False)
orders.to_sql('orders', connection, if_exists='replace', index=False)
products.to_sql('products', connection, if_exists='replace', index=False)
staffs.to_sql('staffs', connection, if_exists='replace', index=False)
stocks.to_sql('stocks', connection, if_exists='replace', index=False)
stores.to_sql('stores', connection, if_exists='replace', index=False)

3

Overview of each table

In [7]:
#Brand Table
query = 'SELECT * FROM  brands'

df = pd.read_sql_query(query, connection)
df

brand_id    brand_name
0         1       Electra
1         2          Haro
2         3        Heller
3         4   Pure Cycles
4         5       Ritchey
5         6       Strider
6         7  Sun Bicycles
7         8         Surly
8         9          Trek

In [11]:
#Brand Table
query = 'SELECT COUNT (*) FROM  brands'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0          9

As we can see, there are 9 rows in the brand table.

In [8]:
#Category Table
query = 'SELECT * FROM  categories'

df = pd.read_sql_query(query, connection)
df

category_id        category_name
0            1    Children Bicycles
1            2     Comfort Bicycles
2            3    Cruisers Bicycles
3            4  Cyclocross Bicycles
4            5       Electric Bikes
5            6       Mountain Bikes
6            7           Road Bikes

In [12]:
#Brand Table
query = 'SELECT COUNT (*) FROM  categories'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0          7

As we can see, there are 7 rows in the categories table.

In [13]:
#Customers Table
query = 'SELECT * FROM  customers LIMIT 5'

df = pd.read_sql_query(query, connection)
df

customer_id  first_name last_name           phone                    email  \
0            1       Debra     Burks            None    debra.burks@yahoo.com   
1            2       Kasha      Todd            None     kasha.todd@yahoo.com   
2            3      Tameka    Fisher            None    tameka.fisher@aol.com   
3            4       Daryl    Spence            None     daryl.spence@aol.com   
4            5  Charolette      Rice  (916) 381-6003  charolette.rice@msn.com   

                  street           city state  zip_code  
0      9273 Thorne Ave.    Orchard Park    NY     14127  
1       910 Vine Street        Campbell    CA     95008  
2  769C Honey Creek St.   Redondo Beach    CA     90278  
3        988 Pearl Lane       Uniondale    NY     11553  
4         107 River Dr.      Sacramento    CA     95820

In [14]:
#Customers Table
query = 'SELECT COUNT (*) FROM  customers'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0       1445

The 'customer' table comprises 9 columns and 1445 rows. Let's examine the number of unique values in certain columns within this table.

In [22]:
#Customers Table
query = 'SELECT DISTINCT first_name FROM  customers'

df = pd.read_sql_query(query, connection)
df

first_name
0          Debra
1          Kasha
2         Tameka
3          Daryl
4     Charolette
...          ...
1260         Son
1261     Florrie
1262      Ernest
1263      Cassie
1264       Ester

[1265 rows x 1 columns]

In [23]:
#Customers Table
query = 'SELECT DISTINCT email FROM  customers'

df = pd.read_sql_query(query, connection)
df

email
0       debra.burks@yahoo.com
1        kasha.todd@yahoo.com
2       tameka.fisher@aol.com
3        daryl.spence@aol.com
4     charolette.rice@msn.com
...                       ...
1440  jamaal.morrison@msn.com
1441   cassie.cline@gmail.com
1442    lezlie.lamb@gmail.com
1443   ivette.estes@gmail.com
1444  ester.acevedo@gmail.com

[1445 rows x 1 columns]

In [24]:
#Customers Table
query = 'SELECT DISTINCT city FROM  customers'

df = pd.read_sql_query(query, connection)
df

city
0       Orchard Park
1           Campbell
2      Redondo Beach
3          Uniondale
4         Sacramento
..               ...
190    Copperas Cove
191  Oakland Gardens
192        Tonawanda
193     Far Rockaway
194         Westbury

[195 rows x 1 columns]

Upon reviewing the outcomes of the three queries mentioned earlier, it becomes apparent that there are 180 individuals with identical names. Therefore, considering an alternative column or merging it with another is necessary for a more in-depth investigation. Conversely, the email column appears to be unique for each customer. Furthermore, a noteworthy observation is that the city information, gleaned from the above queries, reveals 195 distinct cities, which may prove crucial for future analyses.

In [25]:
#Order_items Table
query = 'SELECT * FROM order_items LIMIT 5'

df = pd.read_sql_query(query, connection)
df

order_id  item_id  product_id  quantity  list_price  discount
0         1        1          20         1      599.99      0.20
1         1        2           8         2     1799.99      0.07
2         1        3          10         2     1549.00      0.05
3         1        4          16         2      599.99      0.05
4         1        5           4         1     2899.99      0.20

In [26]:
#Order_items Table
query = 'SELECT COUNT (*) FROM  order_items'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0       4722

In [27]:
#Customers Table
query = 'SELECT DISTINCT discount FROM order_items'

df = pd.read_sql_query(query, connection)
df

discount
0      0.20
1      0.07
2      0.05
3      0.10

The 'order_items' table encompasses 6 columns and consists of 4722 rows. Significantly, it is notable that the provided discounts exhibit variation, ranging between 5%, 7%, 10%, and 20%.

In [29]:
#Orders Table
query = 'SELECT * FROM  orders LIMIT 5'

df = pd.read_sql_query(query, connection)
df

order_id  customer_id  order_status  order_date required_date shipped_date  \
0         1          259             4  2016-01-01    2016-01-03   2016-01-03   
1         2         1212             4  2016-01-01    2016-01-04   2016-01-03   
2         3          523             4  2016-01-02    2016-01-05   2016-01-03   
3         4          175             4  2016-01-03    2016-01-04   2016-01-05   
4         5         1324             4  2016-01-03    2016-01-06   2016-01-06   

   store_id  staff_id  
0         1         2  
1         2         6  
2         2         7  
3         1         3  
4         2         6

In [31]:
#Orders Table
query = 'SELECT COUNT (*) FROM  orders'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0       1615

In [32]:
#Orders Table
query = 'SELECT DISTINCT customer_id FROM orders'

df = pd.read_sql_query(query, connection)
df

customer_id
0             259
1            1212
2             523
3             175
4            1324
...           ...
1440          816
1441          112
1442           35
1443          180
1444          392

[1445 rows x 1 columns]

Upon reviewing the queries for the 'Orders' table above, it's notable that there are 8 columns and a total of 1615 rows. An intriguing observation is that the 'customer_id' column indicates a count fewer than the total rows (1615). Further investigation into this anomaly could provide valuable insights, a task we may delve into later in this notebook.

In [34]:
#Products Table
query = 'SELECT * FROM  products LIMIT 5'

df = pd.read_sql_query(query, connection)
df

product_id                        product_name  brand_id  category_id  \
0           1                     Trek 820 - 2016         9            6   
1           2  Ritchey Timberwolf Frameset - 2016         5            6   
2           3     Surly Wednesday Frameset - 2016         8            6   
3           4            Trek Fuel EX 8 29 - 2016         9            6   
4           5        Heller Shagamaw Frame - 2016         3            6   

   model_year  list_price  
0        2016      379.99  
1        2016      749.99  
2        2016      999.99  
3        2016     2899.99  
4        2016     1320.99

In [35]:
#Products Table
query = 'SELECT COUNT (*) FROM  products'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0        321

In [37]:
#Products Table
query = 'SELECT DISTINCT product_name FROM products'

df = pd.read_sql_query(query, connection)
df

product_name
0                         Trek 820 - 2016
1      Ritchey Timberwolf Frameset - 2016
2         Surly Wednesday Frameset - 2016
3                Trek Fuel EX 8 29 - 2016
4            Heller Shagamaw Frame - 2016
..                                    ...
286          Trek Checkpoint ALR 5 - 2019
287  Trek Checkpoint ALR 5 Women's - 2019
288   Trek Checkpoint SL 5 Women's - 2019
289           Trek Checkpoint SL 6 - 2019
290   Trek Checkpoint ALR Frameset - 2019

[291 rows x 1 columns]

In [39]:
#Products Table
query = 'SELECT DISTINCT model_year FROM products'

df = pd.read_sql_query(query, connection)
df

model_year
0        2016
1        2017
2        2018
3        2019

By observing the above queries  for the 'Products' table, it reveals 6 columns and a total of 321 rows. Among these, there are 291 distinct product names, indicating a diverse product range. Additionally, noteworthy is that the table encompasses data spanning from 2016 to 2019, presenting valuable temporal information for insightful comparisons across different years.

In [40]:
#Staffs Table
query = 'SELECT * FROM  staffs'

df = pd.read_sql_query(query, connection)
df

staff_id  first_name last_name                          email  \
0         1     Fabiola   Jackson     fabiola.jackson@bikes.shop   
1         2      Mireya  Copeland     mireya.copeland@bikes.shop   
2         3       Genna   Serrano       genna.serrano@bikes.shop   
3         4      Virgie   Wiggins      virgie.wiggins@bikes.shop   
4         5    Jannette     David      jannette.david@bikes.shop   
5         6   Marcelene     Boyer     marcelene.boyer@bikes.shop   
6         7      Venita    Daniel       venita.daniel@bikes.shop   
7         8        Kali    Vargas         kali.vargas@bikes.shop   
8         9       Layla   Terrell       layla.terrell@bikes.shop   
9        10  Bernardine   Houston  bernardine.houston@bikes.shop   

            phone  active  store_id  manager_id  
0  (831) 555-5554       1         1         NaN  
1  (831) 555-5555       1         1         1.0  
2  (831) 555-5556       1         1         2.0  
3  (831) 555-5557       1         1         2.0  
4  (516) 379-4444       1         2         1.0  
5  (516) 379-4445       1         2         5.0  
6  (516) 379-4446       1         2         5.0  
7  (972) 530-5555       1         3         1.0  
8  (972) 530-5556       1         3         7.0  
9  (972) 530-5557       1         3         7.0

In [41]:
#Staffs Table
query = 'SELECT COUNT (*) FROM  staffs'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0         10

In [53]:
#Staffs Table
query = 'SELECT DISTINCT staff_id FROM staffs'

df = pd.read_sql_query(query, connection)
df

staff_id
0         1
1         2
2         3
3         4
4         5
5         6
6         7
7         8
8         9
9        10

In [52]:
#Staffs Table
query = 'SELECT store_id, COUNT (staff_id) AS "Staff per Store" FROM staffs GROUP BY store_id'

df = pd.read_sql_query(query, connection)
df

store_id  Staff per Store
0         1                4
1         2                3
2         3                3

Observing the 'staff' table, it consists of 8 columns and comprises 10 rows. Notably, store_id 1 exhibits a higher staff count with 4 members, whereas store_id 2 and 3 each have 3 staff members.

In [54]:
#Stocks Table
query = 'SELECT * FROM  stocks LIMIT 5'

df = pd.read_sql_query(query, connection)
df

store_id  product_id  quantity
0         1           1        27
1         1           2         5
2         1           3         6
3         1           4        23
4         1           5        22

In [56]:
#Staffs Table
query = 'SELECT COUNT (*) FROM  stocks'

df = pd.read_sql_query(query, connection)
df

COUNT (*)
0        939

In [55]:
#Stocks Table
query = 'SELECT DISTINCT product_id FROM stocks'

df = pd.read_sql_query(query, connection)
df

product_id
0             1
1             2
2             3
3             4
4             5
..          ...
308         309
309         310
310         311
311         312
312         313

[313 rows x 1 columns]

In the 'stocks' table, there are only 3 columns and a total of 939 rows. An interesting observation is that, upon examining the unique values in the 'product_id' column, it becomes apparent that out of the 939 total rows, only 313 have unique 'product_id' values.

To be continued...